SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/CrossServerRPCHandler.h"
+
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "SpatialGDKSettings.h"
+
+DEFINE_LOG_CATEGORY(LogCrossServerRPCHandler);
+
+namespace SpatialGDK
+{
+CrossServerRPCHandler::CrossServerRPCHandler(ViewCoordinator& InCoordinator, TUniquePtr<RPCExecutorInterface> InRPCExecutor)
+	: Coordinator(&InCoordinator)
+	, RPCExecutor(MoveTemp(InRPCExecutor))
+{
+}
+
+void CrossServerRPCHandler::ProcessMessages(const TArray<Worker_Op>& WorkerMessages)
+{
+	for (const auto& Op : WorkerMessages)
+	{
+		if (Op.op_type == WORKER_OP_TYPE_COMMAND_REQUEST
+			&& Op.op.command_request.request.command_index == SpatialConstants::UNREAL_RPC_ENDPOINT_COMMAND_ID
+			&& Op.op.command_request.request.component_id == SpatialConstants::SERVER_TO_SERVER_COMMAND_ENDPOINT_COMPONENT_ID)
+		{
+			HandleWorkerOp(Op);
+		}
+	}
+
+	ProcessPendingCrossServerRPCs();
+
+	while (RPCsToDelete.Num() > 0 && RPCsToDelete.HeapTop().Key < FDateTime::Now())
+	{
+		RPCsToDelete.HeapPopDiscard();
+	}
+}
+
+void CrossServerRPCHandler::ProcessPendingCrossServerRPCs()
+{
+	for (auto CrossServerRPCs = QueuedCrossServerRPCs.CreateIterator(); CrossServerRPCs; ++CrossServerRPCs)
+	{
+		int32 ProcessedRPCs = 0;
+		for (const auto& Command : CrossServerRPCs.Value())
+		{
+			if (!TryExecuteCrossServerRPC(Command))
+			{
+				break;
+			}
+
+			RPCGuidsInFlight.Remove(Command.Payload.Id.GetValue());
+			RPCsToDelete.HeapPush(TTuple<FDateTime, uint32>(FDateTime::Now() + Command.TimeoutMillis, Command.Payload.Id.GetValue()));
```

Comment:
Discussed offline - shouldn't use this timeout for defining the lifetime of the guids. If we have some const field for this we can make it easy to tune.